gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java

```diff
@@ -0,0 +1,261 @@
+      SparkSession sparkSession, String catalog, String gcRefName, String gcTable) {
+    this.sparkSession = sparkSession;
+    this.catalogAndTableWithRefName = withRefName(catalog, gcTable, gcRefName);
+    createTableIfAbsent(sparkSession, catalog, TableIdentifier.parse(gcTable), gcRefName);
+  }
+
+  public StructType getSchema() {
+    return schema;
+  }
+
+  /**
+   * Collect the expired contents for the given run id as spark dataset.
+   *
+   * @param runId run id of completed identify task.
+   * @return spark dataset of row where each row is having the expired contents per content id per
+   *     reference.
+   */
+  public Dataset<Row> collectExpiredContentsAsDataSet(String runId) {
+    // collect marker row for the given run-id
+    // Example Query:
+    // SELECT gcRunId FROM nessie.db1.`identified_results@someGcRef`
+    //    WHERE gcRunId = '1a088e65-a6ea-42e9-819d-025f4e22eaec' AND type = 'GC_MARK'
+    Dataset<Row> markerRow =
+        sql(
+            "SELECT %s FROM %s WHERE %s = '%s' AND %s = '%s'",
+            COL_GC_RUN_ID,
+            //
+            catalogAndTableWithRefName,
+            //
+            COL_GC_RUN_ID,
+            runId,
+            //
+            COL_TYPE,
+            TYPE_GC_MARKER);
+    if (markerRow.isEmpty()) {
+      return markerRow;
+    }
+    // Example Query:
+    // SELECT * FROM nessie.db1.`identified_results@someGcRef`
+    //    WHERE gcRunId = '9e809dc9-ac35-41c9-b1c0-5ebcf452ea2d' AND type = 'GC_CONTENT'
+    return sql(
+        "SELECT * FROM %s WHERE %s = '%s' AND %s = '%s'",
+        catalogAndTableWithRefName,
+        //
+        COL_GC_RUN_ID,
+        runId,
+        //
+        COL_TYPE,
+        TYPE_CONTENT);
+  }
+
+  /**
+   * Get the last written run id which is having a marker row.
+   *
+   * @return latest completed run id.
+   */
+  public String getLatestCompletedRunID() {
+    // collect marker row for the last written run id
+    // Example query:
+    // SELECT gcRunId FROM nessie.db1.`identified_results@someGcRef
+    //    WHERE type = 'GC_MARK' ORDER BY gcRunStart DESC LIMIT 1
+    List<Row> markerRows =
+        sql(
+                "SELECT %s FROM %s WHERE %s = '%s' ORDER BY %s DESC LIMIT 1",
+                COL_GC_RUN_ID,
+                //
+                catalogAndTableWithRefName,
+                //
+                COL_TYPE,
+                TYPE_GC_MARKER,
+                //
+                COL_GC_RUN_START)
+            .collectAsList();
+    return markerRows.isEmpty() ? null : markerRows.get(0).getString(0);
+  }
+
+  /**
+   * Convert each task result into a spark row and store it in the output table. Write a marker row
+   * in the end to indicate that the task is completed.
+   *
+   * @param rowDataset content rows to be written.
+   * @param runId run id of current identify task.
+   * @param startedAt task start time.
+   */
+  void writeToOutputTable(Dataset<Row> rowDataset, String runId, Timestamp startedAt) {
```

Comment:
This function writes the results of all tasks and then writes the marker row. Why does it need the marker row then?

Also: why don't the executors write the rows directly but instead pass the rows back to the caller?

So either write the rows from the executors plus the marker row here or don't use the marker row at all.

What are the time + space requirements (the overall algorithm) for both approaches wrt to writing to the output table? Have you tested this approach with 10000 or 100000 tables?

Reply:
> Also: why don't the executors write the rows directly but instead pass the rows back to the caller?

Executor will not have spark context to create the dataframe[1]. Hence, driver collects the row iteratively (not all at once) and writes it to the output table. However overall operation is atomic in nature and no need of marker row. I will remove it.
[1] https://stackoverflow.com/questions/48715661/spark-how-can-i-create-local-dataframe-in-each-executor

> What are the time + space requirements (the overall algorithm) for both approaches wrt to writing to the output table? Have you tested this approach with 10000 or 100000 tables?

It depends on the parallelism, available cores and other configurations. Number of references will be divided into X (available parallelism) and distributively each reference will reverse its commit log. As there is no alternative approach available at the moment, I didn't test performance. Want to have a fully functional working algorithm first and take up the performance optimization in a follow up.

Reply:
say If the number of executor instance is configured as 5,
there will be 5 task launched with an iterator of 2000 references per each task.
5 tasks execute concurrently in finding the expired row per task (that too one task will produce one expired row at a time form Row iterator by working on one reference)
These 5 rows are added to in memory and again next batch of 5 rows are fetched and so on...
As the final action is `.append()`, once the default target file is reached (say 512mb), rows will be flushed to parquet file.

Based on the cpu and memory available, we can tune the executor instance.

Do you have any other suggestion for this?